File: tool/download_assistant/macos/Sources/AssistantCore/Selection.swift
import Foundation

// העתקה נאמנה של tool/release/download_assistant_selection.dart (מימוש הייחוס).
// הבדיקות משוות ל-tool/download_assistant/fixtures/expected-selections.json.

/// גבול הקובץ הבודד: Windows אינו מריץ exe בגודל 4 GiB ומעלה, ו-FAT32 אינו מחזיק קובץ כזה.
public let maxSingleOutputFileSize: Int64 = 4_294_967_296

public let assistantPlatforms = ["windows", "macos", "linux", "android"]

public let platformDisplayNames: [String: String] = [
    "windows": "Windows",
    "macos": "macOS",
    "linux": "Linux",
    "android": "Android",
]

public let portablePackageFormat = "portable"

/// מחשב היעד. ארכיטקטורה ריקה כשלפלטפורמה אין רכיבים תלויי ארכיטקטורה; פורמט ריק מחוץ ל-Linux.
public struct AssistantTarget: Equatable {
    public var platform: String
    public var architecture: String
    public var packageFormat: String

    public init(platform: String, architecture: String = "", packageFormat: String = "") {
        self.platform = platform
        self.architecture = architecture
        self.packageFormat = packageFormat
    }
}

public struct AssistantPreset: Equatable {
    public let id: String
    public let caption: String
    public let description: String
    public let members: [String]
}

private func isWildcard(_ value: String) -> Bool {
    value.isEmpty || value == "any"
}

public func componentFitsTarget(_ component: ManifestComponent, _ target: AssistantTarget) -> Bool {
    if !isWildcard(component.platform) && component.platform != target.platform { return false }
    if !isWildcard(component.architecture) && component.architecture != target.architecture {
        return false
    }
    if !isWildcard(component.packageFormat) && component.packageFormat != target.packageFormat {
        return false
    }
    return true
}

/// Windows מסרב להריץ exe בגודל 4 GiB ומעלה.
private func componentIsRunnable(_ component: ManifestComponent) -> Bool {
    !component.assets.contains {
        $0.name.lowercased().hasSuffix(".exe") && $0.size >= maxSingleOutputFileSize
    }
}

/// המתקין שיתקין את [component] ביעד — הראשון ב-`installedBy` שמוצע בו, או nil.
public func installerFor(
    _ manifest: ReleaseManifest, _ component: ManifestComponent, _ target: AssistantTarget
) -> ManifestComponent? {
    for id in component.installedBy {
        if let installer = manifest.components.first(where: { $0.id == id }),
           componentFitsTarget(installer, target), componentIsRunnable(installer) {
            return installer
        }
    }
    return nil
}

/// מוצע ליעד (בהצעות ובבחירה האישית): מתאים, בלי exe שאי אפשר להריץ, ואם יש לו
/// `installedBy` — אחד ממתקיניו מוצע.
public func componentIsOffered(
    _ manifest: ReleaseManifest, _ component: ManifestComponent, _ target: AssistantTarget
) -> Bool {
    guard componentFitsTarget(component, target), componentIsRunnable(component) else { return false }
    return component.installedBy.isEmpty || installerFor(manifest, component, target) != nil
}

/// הפלטפורמות שיש להן לפחות רכיב ייעודי אחד (רכיב `any` לבדו אינו מספיק).
public func platformChoices(_ manifest: ReleaseManifest) -> [String] {
    let present = Set(manifest.components.map { $0.platform })
    return assistantPlatforms.filter { present.contains($0) }
}

/// x64 תמיד ראשונה; העמוד מוצג רק כשיש יותר מאחת.
public func architectureChoices(_ manifest: ReleaseManifest, _ platform: String) -> [String] {
    var found = Set<String>()
    for component in manifest.components where component.platform == platform {
        if !isWildcard(component.architecture) { found.insert(component.architecture) }
    }
    var sorted = found.sorted()
    if let index = sorted.firstIndex(of: "x64") {
        sorted.remove(at: index)
        sorted.insert("x64", at: 0)
    }
    return sorted
}

/// `portable` מוצע כשיש רכיב תוכנה שאינו תלוי מנהל חבילות.
public func packageFormatChoices(
    _ manifest: ReleaseManifest, _ platform: String, _ architecture: String
) -> [String] {
    var formats = Set<String>()
    var portable = false
    for component in manifest.components where component.platform == platform {
        if !isWildcard(component.architecture) && component.architecture != architecture { continue }
        if !isWildcard(component.packageFormat) {
            formats.insert(component.packageFormat)
        } else if component.type.hasPrefix("application") {
            portable = true
        }
    }
    if formats.isEmpty { return [] }
    return formats.sorted() + (portable ? [portablePackageFormat] : [])
}

private let debFamily: Set<String> = [
    "debian", "ubuntu", "linuxmint", "pop", "elementary", "zorin", "raspbian",
    "kali", "neon", "deepin", "mx",
]

private let rpmFamily: Set<String> = [
    "fedora", "rhel", "centos", "rocky", "almalinux", "ol", "suse", "opensuse",
    "sles", "mageia", "openmandriva", "nobara",
]

/// פורמט ברירת המחדל מתוך `/etc/os-release` (nil = לא ב-Linux). ID נבדק לפני ID_LIKE.
public func defaultPackageFormat(_ osRelease: String?, _ choices: [String]) -> String {
    func pick(_ preferred: String) -> String {
        if choices.contains(preferred) { return preferred }
        return choices.first ?? ""
    }
    guard let osRelease = osRelease else { return pick("deb") }

    var values: [String: String] = [:]
    for line in osRelease.components(separatedBy: "\n") {
        guard let equals = line.firstIndex(of: "="), equals != line.startIndex else { continue }
        var value = String(line[line.index(after: equals)...])
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if value.count >= 2, let first = value.first, first == "\"" || first == "'",
           value.last == first {
            value = String(value.dropFirst().dropLast())
        }
        let key = String(line[..<equals]).trimmingCharacters(in: .whitespacesAndNewlines)
        values[key] = value.lowercased()
    }
    let tokens = [values["ID"] ?? ""]
        + (values["ID_LIKE"] ?? "").split(whereSeparator: { $0.isWhitespace }).map(String.init)
    for token in tokens where !token.isEmpty {
        let family = token.hasPrefix("opensuse") ? "opensuse" : token
        if debFamily.contains(family) { return pick("deb") }
        if rpmFamily.contains(family) { return pick("rpm") }
    }
    return pick(portablePackageFormat)
}

private func collect(
    _ manifest: ReleaseManifest, _ target: AssistantTarget,
    types: Set<String>? = nil, requiredOnly: Bool = false
) -> [String] {
    manifest.components.filter { component in
        componentIsOffered(manifest, component, target)
            && (!requiredOnly || component.required)
            && (types == nil || types!.contains(component.type))
    }.map { $0.id }
}

/// סגירת הבחירה: כל `dependsOn` שמוצע ביעד (שאינו מוצע — מדולג), ולכל רכיב עם
/// `installedBy` שאף מתקין שלו אינו בבחירה — המתקין מ-installerFor. בסדר המניפסט.
public func withDependencies(
    _ manifest: ReleaseManifest, _ members: [String], _ target: AssistantTarget
) -> [String] {
    var byId: [String: ManifestComponent] = [:]
    for component in manifest.components {
        byId[component.id] = component
    }
    var closed = Set(members)
    var changed = true
    while changed {
        changed = false
        for id in Array(closed) {
            guard let component = byId[id] else { continue }
            for dependency in component.dependsOn {
                guard let required = byId[dependency], componentIsOffered(manifest, required, target) else {
                    continue
                }
                if closed.insert(dependency).inserted { changed = true }
            }
            if component.installedBy.isEmpty || component.installedBy.contains(where: { closed.contains($0) }) {
                continue
            }
            if let installer = installerFor(manifest, component, target),
               closed.insert(installer.id).inserted {
                changed = true
            }
        }
    }
    return manifest.components.filter { closed.contains($0.id) }.map { $0.id }
}

/// ההצעות לפי סדר ההצגה. הצעה ריקה, או זהה להצעה קודמת, מושמטת. "בחירה אישית" אינה כאן.
public func buildPresets(_ manifest: ReleaseManifest, _ target: AssistantTarget) -> [AssistantPreset] {
    let components = manifest.components

    var bundle: ManifestComponent?
    for component in components
    where componentIsOffered(manifest, component, target) && component.type == "application-bundle" {
        if bundle == nil || component.downloadSize > bundle!.downloadSize {
            bundle = component
        }
    }

    // בלי חבילה, "מלאה" היא התוכנה עם ספרייה — ובלי ספרייה אין "מלאה".
    let full: [String]
    if let bundle = bundle {
        full = [bundle.id] + components.filter {
            $0.installedBy.contains(bundle.id) && componentIsOffered(manifest, $0, target)
        }.map { $0.id }
    } else {
        let collected = collect(manifest, target, types: ["application", "library", "dependency"])
        let hasLibrary = components.contains { collected.contains($0.id) && $0.type == "library" }
        full = hasLibrary ? collected : []
    }

    let candidates: [(id: String, caption: String, description: String, members: [String])] = [
        (
            "full",
            "התקנה מלאה ומומלצת",
            "התוכנה יחד עם ספריית הספרים — הבחירה המתאימה לרוב המשתמשים.",
            full
        ),
        (
            "basic",
            "התקנה בסיסית (תוכנה בלבד)",
            "התוכנה בלבד. את הספרים אפשר להוריד אחר כך מתוך התוכנה.",
            collect(manifest, target, types: ["application"])
                + collect(manifest, target, requiredOnly: true)
        ),
        (
            "update",
            "עדכון התוכנה בלבד",
            "קובץ ההתקנה של הגרסה החדשה, לעדכון התקנה קיימת.",
            collect(manifest, target, types: ["application"])
        ),
    ]

    var presets: [AssistantPreset] = []
    for candidate in candidates where !candidate.members.isEmpty {
        let closed = withDependencies(manifest, candidate.members, target)
        if closed.isEmpty { continue }
        if presets.contains(where: { $0.members == closed }) { continue }
        presets.append(AssistantPreset(
            id: candidate.id,
            caption: candidate.caption,
            description: candidate.description,
            members: closed
        ))
    }
    return presets
}

/// ל-Windows: רק exe מתחת ל-4 GiB (ארכיון נשאר חלקים — המתקין קורא אותם).
/// לכל יעד אחר: כל נכס מתחת ל-4 GiB, כי שם המשתמש פורס אותו בעצמו.
public func shouldAssembleSplitAsset(_ asset: ManifestAsset, _ targetPlatform: String) -> Bool {
    if asset.size >= maxSingleOutputFileSize { return false }
    if targetPlatform == "windows" {
        return asset.name.lowercased().hasSuffix(".exe")
    }
    return true
}

public func outputSubfolderName(_ targetPlatform: String) -> String {
    "אוצריא להתקנה ל-\(platformDisplayNames[targetPlatform] ?? targetPlatform)"
}

/// הקבצים שייווצרו בתיקיית היעד, בסדר המניפסט.
public func plannedOutputFiles(
    _ manifest: ReleaseManifest, _ selectedIds: [String], _ target: AssistantTarget
) -> [String] {
    let selected = Set(selectedIds)
    var files: [String] = []
    for component in manifest.components where selected.contains(component.id) {
        for asset in component.assets {
            if asset.isSplit && !shouldAssembleSplitAsset(asset, target.platform) {
                files.append(contentsOf: asset.parts.map { $0.name })
            } else {
                files.append(asset.name)
            }
        }
    }
    return files
}

/// '' לקובץ יחיד, אחרת תת-התיקייה.
public func plannedOutputSubfolder(_ files: [String], _ targetPlatform: String) -> String {
    files.count > 1 ? outputSubfolderName(targetPlatform) : ""
}

File: tool/download_assistant/linux/selection.c
#include "selection.h"

#include <string.h>

const char *const otz_assistant_platforms[] = {"windows", "macos", "linux",
                                               "android", NULL};

static const char *const display_names[][2] = {
    {"windows", "Windows"},
    {"macos", "macOS"},
    {"linux", "Linux"},
    {"android", "Android"},
};

static const char *const deb_family[] = {
    "debian", "ubuntu", "linuxmint", "pop",    "elementary", "zorin",
    "raspbian", "kali", "neon",      "deepin", "mx",         NULL};

static const char *const rpm_family[] = {
    "fedora", "rhel", "centos", "rocky",  "almalinux",    "ol",
    "suse",   "opensuse", "sles", "mageia", "openmandriva", "nobara", NULL};

void otz_preset_free(OtzPreset *preset) {
  if (preset == NULL) return;
  g_free(preset->id);
  g_ptr_array_unref(preset->members);
  g_free(preset);
}

const char *otz_platform_display_name(const char *platform) {
  for (gsize i = 0; i < G_N_ELEMENTS(display_names); i++) {
    if (strcmp(display_names[i][0], platform) == 0) return display_names[i][1];
  }
  return platform;
}

gboolean otz_string_array_contains(GPtrArray *array, const char *value) {
  for (guint i = 0; i < array->len; i++) {
    if (strcmp(g_ptr_array_index(array, i), value) == 0) return TRUE;
  }
  return FALSE;
}

static gboolean in_list(const char *const *list, const char *value) {
  for (gsize i = 0; list[i] != NULL; i++) {
    if (strcmp(list[i], value) == 0) return TRUE;
  }
  return FALSE;
}

static gboolean is_wildcard(const char *value) {
  return *value == '\0' || strcmp(value, "any") == 0;
}

static GPtrArray *new_strings(void) {
  return g_ptr_array_new_with_free_func(g_free);
}

gboolean otz_component_fits_target(const OtzComponent *component,
                                   const OtzTarget *target) {
  if (!is_wildcard(component->platform) &&
      strcmp(component->platform, target->platform) != 0)
    return FALSE;
  if (!is_wildcard(component->architecture) &&
      strcmp(component->architecture, target->architecture) != 0)
    return FALSE;
  if (!is_wildcard(component->package_format) &&
      strcmp(component->package_format, target->package_format) != 0)
    return FALSE;
  return TRUE;
}

/* Windows refuses to run an exe of 4 GiB or more. */
static gboolean component_is_runnable(const OtzComponent *component) {
  for (guint a = 0; a < component->assets->len; a++) {
    const OtzAsset *asset = g_ptr_array_index(component->assets, a);
    g_autofree char *lower = g_ascii_strdown(asset->name, -1);
    if (g_str_has_suffix(lower, ".exe") &&
        asset->size >= OTZ_MAX_SINGLE_OUTPUT_FILE_SIZE)
      return FALSE;
  }
  return TRUE;
}

const OtzComponent *otz_installer_for(const OtzManifest *manifest,
                                      const OtzComponent *component,
                                      const OtzTarget *target) {
  for (guint i = 0; i < component->installed_by->len; i++) {
    const OtzComponent *installer = otz_manifest_find(
        manifest, g_ptr_array_index(component->installed_by, i));
    if (installer != NULL && otz_component_fits_target(installer, target) &&
        component_is_runnable(installer))
      return installer;
  }
  return NULL;
}

gboolean otz_component_is_offered(const OtzManifest *manifest,
                                  const OtzComponent *component,
                                  const OtzTarget *target) {
  if (!otz_component_fits_target(component, target) ||
      !component_is_runnable(component))
    return FALSE;
  return component->installed_by->len == 0 ||
         otz_installer_for(manifest, component, target) != NULL;
}

GPtrArray *otz_platform_choices(const OtzManifest *manifest) {
  GPtrArray *choices = new_strings();
  for (gsize p = 0; otz_assistant_platforms[p] != NULL; p++) {
    for (guint i = 0; i < manifest->components->len; i++) {
      const OtzComponent *component = g_ptr_array_index(manifest->components, i);
      if (strcmp(component->platform, otz_assistant_platforms[p]) == 0) {
        g_ptr_array_add(choices, g_strdup(otz_assistant_platforms[p]));
        break;
      }
    }
  }
  return choices;
}

static gint compare_strings(gconstpointer a, gconstpointer b) {
  return strcmp(*(const char *const *)a, *(const char *const *)b);
}

static void add_unique(GPtrArray *array, const char *value) {
  if (!otz_string_array_contains(array, value))
    g_ptr_array_add(array, g_strdup(value));
}

GPtrArray *otz_architecture_choices(const OtzManifest *manifest,
                                    const char *platform) {
  GPtrArray *found = new_strings();
  for (guint i = 0; i < manifest->components->len; i++) {
    const OtzComponent *component = g_ptr_array_index(manifest->components, i);
    if (strcmp(component->platform, platform) != 0) continue;
    if (!is_wildcard(component->architecture))
      add_unique(found, component->architecture);
  }
  g_ptr_array_sort(found, compare_strings);
  for (guint i = 0; i < found->len; i++) {
    if (strcmp(g_ptr_array_index(found, i), "x64") == 0) {
      char *x64 = g_ptr_array_steal_index(found, i);
      g_ptr_array_insert(found, 0, x64);
      break;
    }
  }
  return found;
}

GPtrArray *otz_package_format_choices(const OtzManifest *manifest,
                                      const char *platform,
                                      const char *architecture) {
  GPtrArray *formats = new_strings();
  gboolean portable = FALSE;
  for (guint i = 0; i < manifest->components->len; i++) {
    const OtzComponent *component = g_ptr_array_index(manifest->components, i);
    if (strcmp(component->platform, platform) != 0) continue;
    if (!is_wildcard(component->architecture) &&
        strcmp(component->architecture, architecture) != 0)
      continue;
    if (!is_wildcard(component->package_format))
      add_unique(formats, component->package_format);
    else if (g_str_has_prefix(component->type, "application"))
      portable = TRUE;
  }
  if (formats->len == 0) return formats;
  g_ptr_array_sort(formats, compare_strings);
  if (portable) g_ptr_array_add(formats, g_strdup(OTZ_PORTABLE_PACKAGE_FORMAT));
  return formats;
}

static char *pick_format(GPtrArray *choices, const char *preferred) {
  if (otz_string_array_contains(choices, preferred)) return g_strdup(preferred);
  return g_strdup(choices->len == 0 ? "" : g_ptr_array_index(choices, 0));
}

char *otz_default_package_format(const char *os_release, GPtrArray *choices) {
  if (os_release == NULL) return pick_format(choices, "deb");

  g_autofree char *id = NULL;
  g_autofree char *id_like = NULL;
  g_auto(GStrv) lines = g_strsplit(os_release, "\n", -1);
  for (gsize i = 0; lines[i] != NULL; i++) {
    const char *eq = strchr(lines[i], '=');
    if (eq == NULL || eq == lines[i]) continue;
    g_autofree char *key = g_strstrip(g_strndup(lines[i], eq - lines[i]));
    char *value = g_strstrip(g_strdup(eq + 1));
    gsize length = strlen(value);
    if (length >= 2 && (value[0] == '"' || value[0] == '\'') &&
        value[length - 1] == value[0]) {
      memmove(value, value + 1, length - 2);
      value[length - 2] = '\0';
    }
    char *lower = g_utf8_strdown(value, -1);
    g_free(value);
    if (strcmp(key, "ID") == 0) {
      g_free(id);
      id = lower;
    } else if (strcmp(key, "ID_LIKE") == 0) {
      g_free(id_like);
      id_like = lower;
    } else {
      g_free(lower);
    }
  }

  g_autoptr(GPtrArray) tokens = new_strings();
  if (id != NULL && *id != '\0') g_ptr_array_add(tokens, g_strdup(id));
  if (id_like != NULL) {
    g_auto(GStrv) parts = g_regex_split_simple("\\s+", id_like, 0, 0);
    for (gsize i = 0; parts[i] != NULL; i++) {
      if (*parts[i] != '\0') g_ptr_array_add(tokens, g_strdup(parts[i]));
    }
  }
  for (guint i = 0; i < tokens->len; i++) {
    const char *token = g_ptr_array_index(tokens, i);
    const char *family = g_str_has_prefix(token, "opensuse") ? "opensuse" : token;
    if (in_list(deb_family, family)) return pick_format(choices, "deb");
    if (in_list(rpm_family, family)) return pick_format(choices, "rpm");
  }
  return pick_format(choices, OTZ_PORTABLE_PACKAGE_FORMAT);
}

GPtrArray *otz_with_dependencies(const OtzManifest *manifest,
                                 GPtrArray *members, const OtzTarget *target) {
  g_autoptr(GHashTable) closed = g_hash_table_new(g_str_hash, g_str_equal);
  for (guint i = 0; i < members->len; i++)
    g_hash_table_add(closed, g_ptr_array_index(members, i));
  gboolean changed = TRUE;
  while (changed) {
    changed = FALSE;
    for (guint i = 0; i < manifest->components->len; i++) {
      const OtzComponent *component = g_ptr_array_index(manifest->components, i);
      if (!g_hash_table_contains(closed, component->id)) continue;
      for (guint d = 0; d < component->depends_on->len; d++) {
        const OtzComponent *dependency = otz_manifest_find(
            manifest, g_ptr_array_index(component->depends_on, d));
        if (dependency == NULL ||
            !otz_component_is_offered(manifest, dependency, target))
          continue;
        if (g_hash_table_add(closed, dependency->id)) changed = TRUE;
      }
      gboolean installed = component->installed_by->len == 0;
      for (guint d = 0; !installed && d < component->installed_by->len; d++)
        installed = g_hash_table_contains(
            closed, g_ptr_array_index(component->installed_by, d));
      if (installed) continue;
      const OtzComponent *installer =
          otz_installer_for(manifest, component, target);
      if (installer != NULL && g_hash_table_add(closed, installer->id))
        changed = TRUE;
    }
  }
  GPtrArray *result = new_strings();
  for (guint i = 0; i < manifest->components->len; i++) {
    const OtzComponent *component = g_ptr_array_index(manifest->components, i);
    if (g_hash_table_contains(closed, component->id))
      g_ptr_array_add(result, g_strdup(component->id));
  }
  return result;
}

static gboolean type_in(const OtzComponent *component, const char *const *types) {
  return types == NULL || in_list(types, component->type);
}

static void collect(GPtrArray *out, const OtzManifest *manifest,
                    const OtzTarget *target, const char *const *types,
                    gboolean required_only) {
  for (guint i = 0; i < manifest->components->len; i++) {
    const OtzComponent *component = g_ptr_array_index(manifest->components, i);
    if (otz_component_is_offered(manifest, component, target) &&
        (!required_only || component->required) && type_in(component, types))
      g_ptr_array_add(out, g_strdup(component->id));
  }
}

static gboolean same_list(GPtrArray *a, GPtrArray *b) {
  if (a->len != b->len) return FALSE;
  for (guint i = 0; i < a->len; i++) {
    if (strcmp(g_ptr_array_index(a, i), g_ptr_array_index(b, i)) != 0)
      return FALSE;
  }
  return TRUE;
}

GPtrArray *otz_build_presets(const OtzManifest *manifest,
                             const OtzTarget *target) {
  static const char *const full_types[] = {"application", "library",
                                           "dependency", NULL};
  static const char *const application_types[] = {"application", NULL};
  static const char *const library_types[] = {"library", NULL};

  const OtzComponent *bundle = NULL;
  for (guint i = 0; i < manifest->components->len; i++) {
    const OtzComponent *component = g_ptr_array_index(manifest->components, i);
    if (!otz_component_is_offered(manifest, component, target)) continue;
    if (strcmp(component->type, "application-bundle") != 0) continue;
    if (bundle == NULL || component->download_size > bundle->download_size)
      bundle = component;
  }

  struct {
    const char *id;
    const char *caption;
    const char *description;
    GPtrArray *members;
  } candidates[3] = {
      {"full", "התקנה מלאה ומומלצת",
       "התוכנה יחד עם ספריית הספרים — הבחירה המתאימה לרוב המשתמשים.",
       new_strings()},
      {"basic", "התקנה בסיסית (תוכנה בלבד)",
       "התוכנה בלבד. את הספרים אפשר להוריד אחר כך מתוך התוכנה.", new_strings()},
      {"update", "עדכון התוכנה בלבד",
       "קובץ ההתקנה של הגרסה החדשה, לעדכון התקנה קיימת.", new_strings()},
  };
  if (bundle != NULL) {
    /* The bundle comes with whatever it installs from the folder beside it. */
    g_ptr_array_add(candidates[0].members, g_strdup(bundle->id));
    for (guint i = 0; i < manifest->components->len; i++) {
      const OtzComponent *component = g_ptr_array_index(manifest->components, i);
      if (otz_string_array_contains(component->installed_by, bundle->id) &&
          otz_component_is_offered(manifest, component, target))
        g_ptr_array_add(candidates[0].members, g_strdup(component->id));
    }
  } else {
    g_autoptr(GPtrArray) libraries = new_strings();
    collect(libraries, manifest, target, library_types, FALSE);
    /* Without a library there is no "full" install. */
    if (libraries->len > 0)
      collect(candidates[0].members, manifest, target, full_types, FALSE);
  }
  collect(candidates[1].members, manifest, target, application_types, FALSE);
  collect(candidates[1].members, manifest, target, NULL, TRUE);
  collect(candidates[2].members, manifest, target, application_types, FALSE);

  GPtrArray *presets =
      g_ptr_array_new_with_free_func((GDestroyNotify)otz_preset_free);
  for (gsize c = 0; c < G_N_ELEMENTS(candidates); c++) {
    GPtrArray *members = candidates[c].members;
    GPtrArray *closed = members->len == 0
                            ? NULL
                            : otz_with_dependencies(manifest, members, target);
    g_ptr_array_unref(members);
    if (closed == NULL) continue;
    gboolean duplicate = closed->len == 0;
    for (guint p = 0; !duplicate && p < presets->len; p++) {
      const OtzPreset *earlier = g_ptr_array_index(presets, p);
      duplicate = same_list(earlier->members, closed);
    }
    if (duplicate) {
      g_ptr_array_unref(closed);
      continue;
    }
    OtzPreset *preset = g_new0(OtzPreset, 1);
    preset->id = g_strdup(candidates[c].id);
    preset->caption = candidates[c].caption;
    preset->description = candidates[c].description;
    preset->members = closed;
    g_ptr_array_add(presets, preset);
  }
  return presets;
}

gboolean otz_should_assemble_split_asset(const OtzAsset *asset,
                                         const char *target_platform) {
  if (asset->size >= OTZ_MAX_SINGLE_OUTPUT_FILE_SIZE) return FALSE;
  if (strcmp(target_platform, "windows") == 0) {
    g_autofree char *lower = g_ascii_strdown(asset->name, -1);
    return g_str_has_suffix(lower, ".exe");
  }
  return TRUE;
}

char *otz_output_subfolder_name(const char *target_platform) {
  return g_strconcat("אוצריא להתקנה ל-",
                     otz_platform_display_name(target_platform), NULL);
}

GPtrArray *otz_planned_output_files(const OtzManifest *manifest,
                                    GPtrArray *selected_ids,
                                    const OtzTarget *target) {
  GPtrArray *files = new_strings();
  for (guint i = 0; i < manifest->components->len; i++) {
    const OtzComponent *component = g_ptr_array_index(manifest->components, i);
    if (!otz_string_array_contains(selected_ids, component->id)) continue;
    for (guint a = 0; a < component->assets->len; a++) {
      const OtzAsset *asset = g_ptr_array_index(component->assets, a);
      if (strcmp(asset->kind, "split") == 0 &&
          !otz_should_assemble_split_asset(asset, target->platform)) {
        for (guint p = 0; p < asset->parts->len; p++) {
          const OtzPart *part = g_ptr_array_index(asset->parts, p);
          g_ptr_array_add(files, g_strdup(part->name));
        }
      } else {
        g_ptr_array_add(files, g_strdup(asset->name));
      }
    }
  }
  return files;
}

char *otz_planned_output_subfolder(GPtrArray *files,
                                   const char *target_platform) {
  return files->len > 1 ? otz_output_subfolder_name(target_platform)
                        : g_strdup("");
}

gint64 otz_members_download_size(const OtzManifest *manifest,
                                 GPtrArray *members) {
  gint64 total = 0;
  for (guint i = 0; i < members->len; i++) {
    const OtzComponent *component =
        otz_manifest_find(manifest, g_ptr_array_index(members, i));
    if (component != NULL) total += component->download_size;
  }
  return total;
}
